File: data structures and algorithms/cpp/08_024.cpp
#include <iostream>
#include <vector>
#include <queue>

struct Position {
    int x, y, time;
    Position(int x, int y, int time): 
        x(x), y(y), time(time) {}
};

int floodFillTime(std::vector<std::vector<int>>& maze, 
                 int startX, int startY, 
                 int targetX, int targetY) {
    int rows = maze.size();
    int cols = maze[0].size();
    
    std::vector<std::vector<bool>> visited(
        rows, std::vector<bool>(cols, false));
    
    std::queue<Position> q;
    q.push(Position(startX, startY, 0));
    visited[startX][startY] = true;
    
    std::vector<std::pair<int, int>> directions = 
        {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    
    while (!q.empty()) {
        Position current = q.front();
        q.pop();
        
        if (current.x == targetX && current.y == targetY) {
            return current.time;
        }
        
        for (auto& dir : directions) {
            int nx = current.x + dir.first;
            int ny = current.y + dir.second;
            
            if (nx >= 0 && nx < rows && 
                ny >= 0 && ny < cols && 
                !visited[nx][ny] && maze[nx][ny] == 0) {
                visited[nx][ny] = true;
                q.push(Position(nx, ny, current.time + 1));
            }
        }
    }
    
    return -1;  // Target not reachable
}

int main() {
    std::vector<std::vector<int>> maze = {
        {0, 0, 0, 0},
        {1, 1, 0, 1},
        {0, 2, 0, 0},
        {0, 1, 1, 0}
    };
    
    std::cout << floodFillTime(maze, 2, 1, 0, 0) 
              << std::endl;  // Output: 3
    
    return 0;
}

File: data structures and algorithms/python/08_024.py
from collections import deque

def flood_fill_time(maze, startX, startY, targetX, targetY):
   rows, cols = len(maze), len(maze[0])
   visited = [[False for _ in range(cols)] for _ in range(rows)]
   
   # BFS queue with (x, y, time)
   queue = deque([(startX, startY, 0)])
   visited[startX][startY] = True
   
   directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
   
   while queue:
      x, y, time = queue.popleft()
      
      # If water reaches target
      if x == targetX and y == targetY:
         return time
      
      for dx, dy in directions:
         nx, ny = x + dx, y + dy
         
         # Check if new position is valid
         if (0 <= nx < rows and 0 <= ny < cols and 
             not visited[nx][ny] and maze[nx][ny] == 0):
            visited[nx][ny] = True
            queue.append((nx, ny, time + 1))
   
   return -1

# Test example
maze = [
   [0, 0, 0, 0],
   [1, 1, 0, 1],
   [0, 2, 0, 0],
   [0, 1, 1, 0]
]
print(flood_fill_time(maze, 2, 1, 0, 0))  # Output: 3